tsdb/external.go

```diff
@@ -0,0 +1,22 @@
+package tsdb
+
+import "github.com/prometheus/client_golang/prometheus"
+
+func DBOpts(db *DB) *Options {
+	return db.opts
+}
+
+func DBTimeRetentionCount(db *DB) prometheus.Counter {
+	return db.metrics.timeRetentionCount
+}
+
+func DBSizeRetentionCount(db *DB) prometheus.Counter {
+	return db.metrics.sizeRetentionCount
+}
+
+func DBSetBlocksToDelete(db *DB, blocksToDeleteFactory func(db *DB) BlocksToDeleteFunc) {
+	db.cmtx.Lock()
+	defer db.cmtx.Unlock()
+
+	db.blocksToDelete = blocksToDeleteFactory(db)
```

Comment:
The caller already has a pointer to `DB`, so it is not necessary to capture it within the closure here. This can be done at a higher level, which, in my opinion, would make the code easier to read:
```suggestion
func DBSetBlocksToDelete(db *DB, blocksToDelete BlocksToDeleteFunc) {
	db.cmtx.Lock()
	defer db.cmtx.Unlock()

	db.blocksToDelete = blocksToDelete
```
and instead of factory function:
```go
func PPBlocksToDelete(db *tsdb.DB, dir string, catalog *catalog.Catalog) tsdb.BlocksToDeleteFunc {
	return func(blocks []*tsdb.Block) map[ulid.ULID]struct{} {
		return deletableBlocks(db, dir, catalog, blocks)
	}
}
```

Reply:
we must inject some sort of db factory function, because we use `DB` reference inside `blocksToDelete` closure to calculate prometheus head size